### PR TITLE
gnrc_lorawan: encode LoRaWAN port in netif header

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR encodes the LoRaWAN port in a GNRC netif header.
The benefit of doing this is that we can reuse `txtsnd` and similar commands. From now on it's possible to use GNRC LoRaWAN (almost) out of the box in any GNRC application (e.g `examples/default`).

This PR also deprecates `NETOPT_LORAWAN_TX_PORT` and introduces a `CONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR` to enable this functionality (and don't abruptly change the API of users).

### Testing procedure

The following things should be tested:
- Check that `examples/gnrc_lorawan` works out of the box. The `send` command was replaced by `txtsnd`
- Try sending frames in port 0. Frame shouldn't be sent.
- Try GNRC LoRaWAN with `examples/default`:
```
CFLAGS=-DCONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR USEMODULE=gnrc_lorawan BOARD=b-l072z-lrwan1 make flash term
```
Check that's possible to communicate.

### Issues/PRs references
None so far

### CODE REVIEW

**Review by leandrolanzieri — CHANGES_REQUESTED** [review]

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c [hunk 1]
```diff
@@ -89,15 +90,40 @@ static inline void _set_be_addr(gnrc_lorawan_t *mac, uint8_t *be_addr)
...
+                     ? GNRC_NETTYPE_UNDEF
+                     : GNRC_NETTYPE_LORAWAN;
+    uint32_t demux = IS_ACTIVE(CONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR)
+                     ? GNRC_NETREG_DEMUX_CTX_ALL
+                     : ind->data.port;
+
     gnrc_pktsnip_t *pkt = gnrc_pktbuf_add(NULL, ind->data.pkt->iol_base,
                                           ind->data.pkt->iol_len,
-                                          GNRC_NETTYPE_LORAWAN);
+                                          nettype);
```
[the remark this answers is not in the document]
Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:103]: I don't know how I missed that one

[the remark this answers is not in the document]
Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:103]: fixed in https://github.com/RIOT-OS/RIOT/pull/16110

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c [hunk 2]
```diff
@@ -89,15 +90,40 @@ static inline void _set_be_addr(gnrc_lorawan_t *mac, uint8_t *be_addr)
 
 void gnrc_lorawan_mcps_indication(gnrc_lorawan_t *mac, mcps_indication_t *ind)
 {
-    (void)mac;
+    gnrc_netif_t *netif = container_of(mac, gnrc_netif_t, lorawan.mac);
+    gnrc_nettype_t nettype = IS_ACTIVE(CONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR)
+                     ? GNRC_NETTYPE_UNDEF
+                     : GNRC_NETTYPE_LORAWAN;
+    uint32_t demux = IS_ACTIVE(CONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR)
```
Comment (akshaim) [inline sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: Please assert (port == 0 || port >= 224).In future we may need to accommodate for Mac compliance test scenarios (224) as well.

File: examples/gnrc_lorawan/README.md [hunk 3]
```diff
@@ -124,18 +128,34 @@ E.g send confirmable messages:
...
+TTN, this can be done under `Applications > <APP> > Devices > <DEV> > Overview`
+and then check the `Downlink` section.
+
+After sending data, the LoRaWAN Network Server will reply with the downlink
+data. For simplicity, this application is configured to dump downlink data to
+[GNRC pktdump](https://doc.riot-os.org/pktdump_8h.html).
+
+E.g:
+```
+
```
Comment (jia200x) [inline examples/gnrc_lorawan/README.md:153]: note to self: please add example

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c (hunk 2 above)
Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: so, basically you add a remote DoS depending of the content of the downlink? (:

Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: I can add port 0 though, since that shouldn't go up to the stack

Comment (akshaim) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: As per 1.0.3 spec, 1-224 is the only allowed range of values for the port. We used to enforce it in the application code previously.
Also, in the current scenario, values beyond 255 are also accepted which should return an error, right ?

Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: Wait, I just noticed that the LORAMAC constants are already defining this values (at least for reception). So this means, we can rely on that and assume no frame will be sent out of the valid "application specific" range (1-223). However, reception would need to adapt a similar logic.

> Also, in the current scenario, values beyond 255 are also accepted which should return an error, right ?

Considering that the `port` is an 8 bit value, I don't think that would happen

Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: done!

File: examples/gnrc_lorawan/README.md (hunk 3 above)
Comment (jia200x) [reply examples/gnrc_lorawan/README.md]: done!

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jia200x) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:105]: done!

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c (hunk 2 above)
Comment (akshaim) [reply sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:97]: Great :)

**Comment by jia200x** [comment]
rebased!

**Comment by MrKevinWeiss** [comment]
Also did a quick test:

```
Welcome to pyterm!
Type '/exit' to exit.
version
2021-06-14 15:57:15,777 # version
2021-06-14 15:57:15,783 # OPENTHREAD/thread-reference-20180926; NONE; Jun 14 2021 15:54:31
2021-06-14 15:57:15,784 # Done
> panid
2021-06-14 15:57:17,977 #  panid
2021-06-14 15:57:17,978 # beef
2021-06-14 15:57:17,978 # Done
> 
```

**Comment by MrKevinWeiss** [comment]
I was able to run on my board with credentials from @jia200x:

<details><summary><code>CFLAGS=-DCONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR USEMODULE='gnrc gnrc_netdev_default auto_init_gnrc_netif gnrc_txtsnd gnrc_pktdump gnrc_lorawan sx1272' DRIVER=sx1272 BOARD=nucleo-l452re make flash term -C examples/default/</code></summary>

```
2021-06-14 16:29:27,977 # ifconfig 3 up
txtsnd 05 whateverIwant
2021-06-14 16:29:37,458 # txtsnd 05 whateverIwant
2021-06-14 16:29:37,462 # usage: txtsnd <if> [<L2 addr>|bcast] <data>
txtsnd 3 01 whateverIwant
2021-06-14 16:29:46,082 # txtsnd 3 01 whateverIwant
txtsnd 3 02 whateverIwant
2021-06-14 16:30:03,595 # txtsnd 3 02 whateverIwant
> 2021-06-14 16:30:07,415 # PKTDUMP: data received:
2021-06-14 16:30:07,419 # ~~ SNIP  0 - size:   4 byte, type: NETTYPE_UNDEF (0)
2021-06-14 16:30:07,421 # 00000000  DE  AD  BE  ED
2021-06-14 16:30:07,426 # ~~ SNIP  1 - size:   9 byte, type: NETTYPE_NETIF (-1)
2021-06-14 16:30:07,429 # if_pid: 3  rssi: -32768  lqi: 0
2021-06-14 16:30:07,429 # flags: 0x0
2021-06-14 16:30:07,431 # src_l2addr: (nil)
2021-06-14 16:30:07,433 # dst_l2addr: 02
2021-06-14 16:30:07,436 # ~~ PKT    -  2 snips, total size:  13 byte
> ifconfig
2021-06-14 16:43:07,631 # ifconfig
2021-06-14 16:43:07,641 # Iface  3  HWaddr: 26:01:5C:73  Frequency: 869524963Hz  RSSI: 117  BW: 125kHz  SF: 9  CR: 4/5  Link: up 
2021-06-14 16:43:07,648 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 0  Num gateways.: 0 
2021-06-14 16:43:07,650 #           IQ_INVERT  
2021-06-14 16:43:07,653 #           RX_SINGLE  OTAA  L2-PDU:255
```
</details>

Note, the `BOARD_PROVIDES_NETIF` cannot be overridden so I had to manually add all the modules since I have using a shield on a generic nucleo.

**Comment by MrKevinWeiss** [comment]
I also went through the documentation and played around with it
```
USEMODULE=sx1272 DRIVER=sx1272 BOARD=nucleo-l452re make flash term -C examples/gnrc_lorawan/
```

<details><summary>device</summary>

```
weiss@mobiweiss:~/wd/RIOT$ USEMODULE=sx1272 DRIVER=sx1272 BOARD=nucleo-l452re make flash term -C examples/gnrc_lorawan/
make: Entering directory '/home/weiss/wd/RIOT/examples/gnrc_lorawan'
Building application "gnrc_lorawan" for "nucleo-l452re" with MCU "stm32".

"make" -C /home/weiss/wd/RIOT/boards/nucleo-l452re
"make" -C /home/weiss/wd/RIOT/boards/common/nucleo
"make" -C /home/weiss/wd/RIOT/core
"make" -C /home/weiss/wd/RIOT/cpu/stm32
"make" -C /home/weiss/wd/RIOT/cpu/cortexm_common
"make" -C /home/weiss/wd/RIOT/cpu/cortexm_common/periph
"make" -C /home/weiss/wd/RIOT/cpu/stm32/periph
"make" -C /home/weiss/wd/RIOT/cpu/stm32/stmclk
"make" -C /home/weiss/wd/RIOT/cpu/stm32/vectors
"make" -C /home/weiss/wd/RIOT/drivers
"make" -C /home/weiss/wd/RIOT/drivers/netdev
"make" -C /home/weiss/wd/RIOT/drivers/periph_common
"make" -C /home/weiss/wd/RIOT/drivers/sx127x
"make" -C /home/weiss/wd/RIOT/sys
"make" -C /home/weiss/wd/RIOT/sys/auto_init
"make" -C /home/weiss/wd/RIOT/sys/crypto
"make" -C /home/weiss/wd/RIOT/sys/div
"make" -C /home/weiss/wd/RIOT/sys/fmt
"make" -C /home/weiss/wd/RIOT/sys/frac
"make" -C /home/weiss/wd/RIOT/sys/hashes
"make" -C /home/weiss/wd/RIOT/sys/iolist
"make" -C /home/weiss/wd/RIOT/sys/isrpipe
"make" -C /home/weiss/wd/RIOT/sys/luid
"make" -C /home/weiss/wd/RIOT/sys/malloc_thread_safe
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/link_layer/lorawan
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/netapi
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/netif
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/netif/hdr
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/netif/init_devs
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/netif/lorawan
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/netreg
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/pkt
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/pktbuf
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/pktbuf_static
"make" -C /home/weiss/wd/RIOT/sys/net/gnrc/pktdump
"make" -C /home/weiss/wd/RIOT/sys/net/link_layer/eui_provider
"make" -C /home/weiss/wd/RIOT/sys/net/link_layer/l2util
"make" -C /home/weiss/wd/RIOT/sys/net/netif
"make" -C /home/weiss/wd/RIOT/sys/newlib_syscalls_default
"make" -C /home/weiss/wd/RIOT/sys/od
"make" -C /home/weiss/wd/RIOT/sys/pm_layered
"make" -C /home/weiss/wd/RIOT/sys/random
"make" -C /home/weiss/wd/RIOT/sys/random/tinymt32
"make" -C /home/weiss/wd/RIOT/sys/shell
"make" -C /home/weiss/wd/RIOT/sys/shell/commands
"make" -C /home/weiss/wd/RIOT/sys/stdio_uart
"make" -C /home/weiss/wd/RIOT/sys/tsrb
"make" -C /home/weiss/wd/RIOT/sys/xtimer
"make" -C /home/weiss/wd/RIOT/sys/ztimer
   text	   data	    bss	    dec	    hex	filename
  54144	    196	   6472	  60812	   ed8c	/home/weiss/wd/RIOT/examples/gnrc_lorawan/bin/nucleo-l452re/gnrc_lorawan.elf
/home/weiss/wd/RIOT/dist/tools/openocd/openocd.sh flash /home/weiss/wd/RIOT/examples/gnrc_lorawan/bin/nucleo-l452re/gnrc_lorawan.elf
### Flashing Target ###
Open On-Chip Debugger 0.11.0+dev-00035-g8d6f7c922-dirty (2021-03-16-15:33)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : DEPRECATED target event trace-config
srst_only separate srst_nogate srst_open_drain connect_assert_srst

Info : clock speed 500 kHz
Info : STLINK V2J29M18 (API v2) VID:PID 0483:374B
Info : Target voltage: 3.247800
Info : stm32l4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for stm32l4x.cpu on 0
Info : Listening on port 40869 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32l4x.cpu       hla_target little stm32l4x.cpu       reset

Info : Unable to match requested speed 500 kHz, using 480 kHz
Info : Unable to match requested speed 500 kHz, using 480 kHz
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08000c50 msp: 0x20000200
Info : device idcode = 0x20016462 (STM32L45/L46xx - Rev Y : 0x2001)
Info : flash size = 512kbytes
Info : flash mode : single-bank
Info : Padding image section 1 at 0x0800d444 with 4 bytes (bank write end alignment)
Warn : Adding extra erase range, 0x0800d448 .. 0x0800d7ff
auto erase enabled
wrote 54344 bytes from file /home/weiss/wd/RIOT/examples/gnrc_lorawan/bin/nucleo-l452re/gnrc_lorawan.elf in 2.447983s (21.679 KiB/s)

verified 54340 bytes in 1.679086s (31.604 KiB/s)

Info : Unable to match requested speed 500 kHz, using 480 kHz
Info : Unable to match requested speed 500 kHz, using 480 kHz
shutdown command invoked
Done flashing
/home/weiss/wd/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200"  
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2021-06-15 10:04:59,602 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
ifconfig 3 set deveui ****
2021-06-15 10:05:08,276 # ifconfig 3 set deveui ****
2021-06-15 10:05:08,282 # success: set long address on interface 3 to ****
> ifconf 3 set appeui ****
2021-06-15 10:05:12,332 # ifconfig 3 set appeui ****
2021-06-15 10:05:12,337 # success: set AppEUI on interface 3 to ****
> ifconf 3 set appkey ****
2021-06-15 10:05:15,134 # ifconfig 3 set appkey ****
2021-06-15 10:05:15,139 # success: set AppKey on interface 3 to *****
> ifconfig up
2021-06-15 10:05:17,129 # ifconfig 3 up
ifconfig
2021-06-15 10:05:39,385 # ifconfig
2021-06-15 10:05:39,394 # Iface  3  HWaddr: 26:01:4B:CA  Frequency: 869524963Hz  RSSI: 117  BW: 125kHz  SF: 12  CR: 4/5  Link: up 
2021-06-15 10:05:39,401 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 0  Num gateways.: 0 
2021-06-15 10:05:39,403 #           IQ_INVERT  
2021-06-15 10:05:39,407 #           RX_SINGLE  OTAA  L2-PDU:255  
2021-06-15 10:05:39,408 #           
> txtsnd 3 7B "Hello RIOT!"
2021-06-15 10:06:00,146 # txtsnd 3 7B "Hello RIOT!"
> 2021-06-15 10:06:03,844 # PKTDUMP: data received:
2021-06-15 10:06:03,849 # ~~ SNIP  0 - size:  12 byte, type: NETTYPE_UNDEF (0)
2021-06-15 10:06:03,854 # 00000000  00  65  11  11  01  65  22  22  02  65  33  33
2021-06-15 10:06:03,859 # ~~ SNIP  1 - size:   9 byte, type: NETTYPE_NETIF (-1)
2021-06-15 10:06:03,861 # if_pid: 3  rssi: -32768  lqi: 0
2021-06-15 10:06:03,862 # flags: 0x0
2021-06-15 10:06:03,863 # src_l2addr: (nil)
2021-06-15 10:06:03,865 # dst_l2addr: 01
2021-06-15 10:06:03,869 # ~~ PKT    -  2 snips, total size:  21 byte
ifconfig 3 set dr 2
2021-06-15 10:06:09,770 # ifconfig 3 set dr 2
2021-06-15 10:06:09,774 # success: set datarate on interface 3 to 2
> ifconfig 3 link_check
2021-06-15 10:06:17,546 # ifconfig 3 link_check
2021-06-15 10:06:17,548 # success: set option
> txtsnd 3 01 "Join the RIOT!"
2021-06-15 10:06:28,251 # txtsnd 3 01 "Join the RIOT!"
> ifconfig 3 ack_req
2021-06-15 10:06:40,114 # ifconfig 3 ack_req
2021-06-15 10:06:40,116 # success: set option
> txtsnd 3 01 "My confirmable message"
2021-06-15 10:06:47,956 # txtsnd 3 01 "My confirmable message"
> txtsnd 3 01 "My confirmable message"
2021-06-15 10:07:01,212 # txtsnd 3 01 "My confirmable message"
> txtsnd 3 01 "My confirmable message"
2021-06-15 10:07:05,979 # txtsnd 3 01 "My confirmable message"
> 
```
</details>

<details><summary>application interface</summary>

Screenshot from 2021-06-15 10-14-16

</details>

**Comment by MrKevinWeiss** [comment]
If you can just fix the cpp check issues, squash rebase all that fun stuff I think we should be OK.

**Comment by jia200x** [comment]
addressed and amended directly

**Review by MrKevinWeiss — APPROVED** [review]
Tested, nice cleanup, ACK.

File: sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c [hunk 4]
```diff
@@ -251,23 +280,51 @@ static int _send(gnrc_netif_t *netif, gnrc_pktsnip_t *payload)
...
+    uint8_t port;
+    int res = -EINVAL;
+
+    if (IS_ACTIVE(CONFIG_GNRC_NETIF_LORAWAN_NETIF_HDR)) {
+        gnrc_netif_hdr_t *netif_hdr;
+        const uint8_t *dst;
+        netif_hdr = payload->data;
+        dst = gnrc_netif_hdr_get_dst_addr(netif_hdr);
+
+        assert(payload->type = GNRC_NETTYPE_NETIF);
```
Comment (leandrolanzieri) [inline sys/net/gnrc/netif/lorawan/gnrc_netif_lorawan.c:293]:
```suggestion
        assert(payload->type == GNRC_NETTYPE_NETIF);
```

File: examples/gnrc_lorawan/Makefile [hunk 5]
```diff
@@ -33,6 +42,14 @@ include $(RIOTBASE)/Makefile.include
 # Check if being configured via Kconfig
 ifndef CONFIG_KCONFIG_USEMODULE_LORAWAN
 
+  # Tell GNRC to encode LoRaWAN port in the GNRC netif header.
+  # This allows us to use `gnrc_txtsnd` to send data from the shell using the
+  # `txtsnd` command.
+  #
+  # Note: From Release 21.10 all GNRC LoRaWAN packets will include the netif
```
Comment (leandrolanzieri) [inline examples/gnrc_lorawan/Makefile:49]: This should be 22.01 right?

**Comment by leandrolanzieri** [comment]
Please squash @jia200x !

**Review by leandrolanzieri — APPROVED** [review]
ACK!

**Pull request merged**